### ISSUE TITLE
dtrace probes emit byte arrays for keys, not null terminated strings

### ISSUE THREAD

**Issue by dalehamel** [memcached/memcached#576]
# TL;DR

After some discussion, it seems the various dtrace commands implemented in memcached presently return a void *, not actually a char * by conventional standards. This means that it may be arbitrary binary data, not necessarily a null terminated string, and you may read more data for the key than intended.

# Why is this a problem?

Reading this data with eBPF right now, the `keylen` will be shorter than the text actually read. This is because the entire buffer is returned, containing extra data from the adjacent struct field members. This is problematic when loading the data into an eBPF map for performance analysis, keys cannot directly correspond to eBPF map keys. While this can be cleaned up, it leads to a less efficient sampling of memcached.

# Solution

## bpftrace

Use the `nkey` argument from dtrace probes to read only n bytes, eg, use `str($buf, $len)` in bpftrace. 

## bcc

Use the `&` operator (or other workaround) to prove to the eBPF verifier that the USDT argument read is less than the size of the buffer. If the buffer size is 0xff or a similar sentinal value, it is easy to prove this with a bitwise comparison.

# Original issue

I've been working on instrumenting memcached using iovisor/bcc and building memcached with `--enable-dtrace`, in order to build performance analysis tools for memcached using USDT probes on linux.

Ordinarily, data for arguments is read like this https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#6-usdt-probes

You can also use https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#2-bpf_probe_read_str, which is what I'm using. However, it assumes that the key is a null terminated string.

After scratching my head a lot, trying to figure out why the keys I were pulling out from my sample script had payload data seemingly arbitrarily tacked onto them, I started to look into why this might be happening in memcached. 

The probe I am trying to use is for `command__set` : https://github.com/memcached/memcached/blob/master/memcached_dtrace.d#L205-L214

It is called here: https://github.com/memcached/memcached/blob/e7c7a1089c066d105bf0a6c1a812668644a4a6f0/memcached.c#L1378

And you can see that the from the definition of `ITEM_key`, it is just calculating the pointer address for the start of the key data: https://github.com/memcached/memcached/blob/e7c7a1089c066d105bf0a6c1a812668644a4a6f0/memcached.h#L116

However, it makes no attempt to actually limit the bytes sent to `it->nkey`, though it **does** pass this to the dtrace probe as `keylen`.

Looking elsewhere in the memcached source, I see this code, with a comment indicating exactly what I suspected - that the key data is not / may not be stored with a null terminator:
https://github.com/memcached/memcached/blob/e7c7a1089c066d105bf0a6c1a812668644a4a6f0/items.c#L621-L622

Luckily I am able to use the `keylen` argument passed to fix up the data in userspace after i have read it, but this isn't ideal. To reproduce the issue, I am able to use bpftrace from ubuntu eoan with the following bpftrace script (very similar to dtrace scripts):

```

BEGIN
{
  printf("%-20s %10s %10s %10s\n", "MEMCACHED KEY", "CALLS", "OBJSIZE", "REQ/s");

  @start = nsecs;
}

usdt::memcached:command__set
{
  @calls[str(arg1)]++;

  $objsize = arg3;
  $interval = (nsecs - @start) / 1000000000;
  $cps = @calls[str(arg1)] / $interval;
  printf("%-20s %10d, %10d %10d\n", str(arg1), @calls[str(arg1)],
                               $objsize, $cps)
}

END
{
  clear(@start);
  clear(@calls);
}

```

Pointing this at a memached process with dtrace probes enabled, and running a benchmark tool like memtier_benchmark:

```
memtier_benchmark --server localhost --port 11211 -P memcache_text
```

I see whet I print the key that it bleeds a bunch of the data out with it.

Would it a feasible workaround for this be to do a similar strncpy of the key to a temporary buffer before emitting the data to the dtrace probe? There are some moderate performance implications to doing so, but I think they are justified.

**Comment by dormando** [comment]
Why can't the probe use the keylen that was passed to it?

**Comment by dalehamel** [comment]
@dormando in eBPF it doesn’t seem possible to do a memcpy or read of an arbitrary size, because the verifier can’t assert it won’t be out of bounds.

I am able to work around this issue in the a python version of the above script by using the keylen value as indicated, but it results in a much less efficient usage. The values are stored in eBPF maps using their mangled keys, read into userspace, and then clamped to the right size using keylen.

In bpftrace however there is no way to do this, as it can’t slice the string. Ideally, the data emitted by memcached be null terminated so that the bpf functions that copy the data know where to stop, as they negates the need for any of these workarounds.

**Comment by dormando** [comment]
This seems a little odd to me; how do you look at arbitrary binary data in a probe?

**Comment by dalehamel** [comment]
>This seems a little odd to me

can you elaborate?

> how do you look at arbitrary binary data in a probe?

https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#1-bpf_probe_read

This is the typical way, if a reading a dtrace/usdt argument returns a pointer you can use this `bpf_probe_read` to get the data, as many examples do. You can see that it accepts a length to read, however, this is typically set to the size of the container, eg other tools by Brendan Gregg (https://github.com/iovisor/bcc/blob/master/tools/mysqld_qslower.py#L84*) or Sasha Goldshtein (https://github.com/iovisor/bcc/blob/master/tools/lib/ucalls.py#L161-L163)

To avoid reading more bytes than you want/need to if you are reading strings, the form that I mentioned `bpf_probe_read_str` will read until it hits a null terminator - so it is the ideal helper for copying string data. This is the helper that I am trying to use.

**Comment by dormando** [comment]
Sorry, OOTL on latest and greatest probing. just looked at the kernel source for these probes. The issue with using the str read is keys may not always be ascii-clean. It's currently possible for them not to be with the binary protocol, but most clients still disallow binary keys... The cachedump command you quoted is actually wrong here; but that's a deprecated command anyway. Keys are treated like arbitrary binary elsewhere (crawler.c, various binprot verbose dumpers)

https://github.com/iovisor/bcc/issues/1260#issuecomment-406365168
^ seems like you can give some non-const sizes to the verifier so long as you verify nkey vs the buffer size (but nkey is 8bit so maybe that's easier to do here?)

**Comment by dalehamel** [comment]
> The issue with using the str read is keys may not always be ascii-clean. It's currently possible for them not to be with the binary protocol, but most clients still disallow binary keys... 

Ah, thank you for pointing this out, that will be useful to know in designing the tool I am working on.

> iovisor/bcc#1260 (comment)  seems like you can give some non-const sizes to the verifier so long as you verify nkey vs the buffer size (but nkey is 8bit so maybe that's easier to do here?)

thanks for the deep dive on this @dormando, I tried to revise my code as follows:

```
...
     u64 keystr = 0;
     int32_t bytecount = 0;
     uint8_t keysize = 0;

     bpf_usdt_readarg(2, ctx, &keystr);
     bpf_usdt_readarg(3, ctx, &keysize);
     bpf_usdt_readarg(4, ctx, &bytecount);

     uint8_t readsize = keysize < (sizeof(keyhit.keystr)-1) ? keysize : sizeof(keyhit.keystr);

     bpf_probe_read(&keyhit.keystr, readsize, (void *)keystr);
...
```

But it still throws the verifier error indicated in that issue:

```
invalid indirect read from stack off -256+250 size 255
```

Which does not occur if a const size is used. I'll keep pursuing this avenue though, hopefully I'll find a suitable workaround.

**Comment by dormando** [comment]
Where's that 250 coming from?

I don't have the background to deal with the error, sorry. I'd have to go read the same stuff you would :) It should be possible with some tweaks... read_str could just as easily read too far if whatever you told it to read happens to not be NULL'ed, so no matter what it needs to be safe to read up to your probe's buffer in size?

ie; internally memory is slabbed in 1MB pages, then chunked out to class size. nkey can be up to 250, but the smallest class sizes are ~96b or so whatever chunk is at the end of the page has the same safety issues regardless of probe_read or probe_read_str, which I would assume the kernel knows how to deal with.

**Comment by dalehamel** [comment]
> Where's that 250 coming from? ... I don't have the background to deal with the error, sorry. I'd have to go read the same stuff you would

Yeah, I didn't mean to suggest as much, just showing that is the same sort of error from https://github.com/iovisor/bcc/issues/1260#issuecomment-406365168 - thanks for investigating as much as you have!

I'll keep digging, thanks for all the leads here and ideas here.

**Comment by dormando** [comment]
Good luck! Please update the issue if you find an answer, or if I should take another look. People find these via google frequently. The probes may also have bitrotted since I don't really test them myself and haven't added new ones in a long time :/ Been more focused on the logger system.

**Comment by dalehamel** [comment]
> Good luck! Please update the issue if you find an answer, or if I should take another look. People find these via google frequently

Absolutely, and I intend to reference this issue and bcc#1260 when I submit my tool, hopefully that will help to find a more elegant solution. For the time being, I have a sufficient workaround for me tool to work - just not as elegantly as I had hoped.

> The probes may also have bitrotted since I don't really test them myself and haven't added new ones in a long time

Unsurprising, I suspect this is because dtrace had been somewhat niche for praduction systems, but my hope is that there will be a 'new wind' behind this standard now that support is rolling out on linux, and the barrier to entry / learning curve is being lowered substantially :)

**Comment by dalehamel** [comment]
> People find these via google frequently

I updated the issue description with a tl;dr and some other details of the issue, should anyone else run into this.

**Comment by dalehamel** [comment]
Following up, with bpftrace this problem can be worked around pretty easily, at least for string keys.

Looking at the `str` builtin it actually does accept a second argument for length and uses a similar process to determine the variable length read:

https://github.com/iovisor/bpftrace/blob/master/src/ast/codegen_llvm.cpp#L413-L441

So in my sample script, replacing `str(arg1)` with `str(arg1, arg2)` i am able to correctly read the values from memcached.

The issue appears to be that bcc does its code generation from C slightly differently, and i've not yet figured out the incantations to produce similar/equivalent bpf code as bpftrace, in order to satisfy the verifier.

**Comment by dalehamel** [comment]
To work around the issue in BCC, I worked off of two assumptions. From memcached source:

https://github.com/memcached/memcached/blob/e7c7a1089c066d105bf0a6c1a812668644a4a6f0/memcached.h#L39-L40

I can see that the max key size is darn close to 2^8=256, meaning it'll fit in a uint8. That's means that in my bcc script, I can set the max buffer size to 255 by just using 0xff, and then before the probe read I can mask the actual `keysize` this sentinal value to use the `&` bitwise and operator to convince the eBPF verifier that the key cannot be possibly out of bounds, per https://git.kernel.org/pub/scm/linux/kernel/git/davem/net-next.git/commit/kernel/bpf/verifier.c?id=06c1c049721a995dee2829ad13b24aaf5d7c5cce

Thanks again for the help @dormando, I'd say this is actually more or less solved for memcached, so
long as the key size is something that is easy to hex mask like this. Referring me to that bcc issue
is what lead me to this linux commit comment to work around the issue.

**Comment by dormando** [comment]
Cool! Congrats. I tried to hint at that earlier in the thread (nkey being 8b), should've made that more assertive :)

I'll leave this open a while in case I want to link to it from a wiki page or something.
